crates/c-api/include/wasi.h

```diff
@@ -156,6 +225,36 @@ WASI_API_EXTERN void wasi_config_inherit_stderr(wasi_config_t* config);
  */
 WASI_API_EXTERN bool wasi_config_preopen_dir(wasi_config_t* config, const char* path, const char* guest_path);
 
+/**
+ * \brief Creates a new empty pipe pair.
+ *
+ * `limit` is maximum pipe size. Use `SIZE_MAX` for practically unbounded pipe.
+ * 
+ * The caller is expected to deallocate both returned pipes.
+ */
+WASI_API_EXTERN void wasi_pipe_new(size_t limit, own wasi_read_pipe_t** ret_read_pipe, own wasi_write_pipe_t** ret_write_pipe);
+
+/**
+ * \brief Returns the number of bytes readable from a given pipe.
+ */
+WASM_API_EXTERN size_t wasi_read_pipe_len(const wasi_read_pipe_t* read_pipe);
+
+/**
+ * \brief Reads from a pipe into a memory slice.
+ * 
+ * `buf` is a memory slice of `buf_len` bytes.
+ * Returns the number of bytes written to `buf`. 0 if the pipe is empty.
+ */
+WASM_API_EXTERN size_t wasi_read_pipe_read(wasi_read_pipe_t* read_pipe, byte_t* buf, size_t buf_len);
```

Comment:
I think it's ok to use `void*` instead of `byte_t*` to be more akin to `read`/`write`